DOC: improve plot_compare_calibration.py

ogrisel opened the pull request:
Related to the ongoing discussion on improving the user guide section that is based on this example: #28171.

I realized that using the default value of C=1 was yielding under-confident predictions, contrary to what was said in the user guide and the text of the example. I decided to tune the value of C using `LogisticRegressionCV` based on a proper scoring rule (I tried both Brier score and nll and both find the same level for C). It seems to lead to better calibrated logistic regression models for most dataset seeds. I think it's a good practice therefore decide to update the example accordingly.

I also noticed that some things we explained in the text of the example would not hold for other choices of the dataset seed (or other dataset parameters that should not matter). So I revised the text accordingly.

ogrisel committed:
DOC: improve plot_compare_calibration.py

ogrisel commented:
With the tuned C, the LR model is slightly under-confident while it was over-confident previously.

Still it looks like an improvement to me.

ogrisel committed:
Iterate on calibration of LR models

thomasjpfan commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,48 +156,84 @@ def predict_proba(self, X):
+# regularized or if they are mis-specified (e.g. the true decision boundary is
+# a highly non-linear function of the input features). In this example with a
+# very small training set size the calibration curve of the logistic regression
+# model is closest but not perfectly on the diagonal: this model is slightly
+# under-confident.
```
From the calibration plot, logistic regression is slightly under confident on the right side and slightly over confident on the left side.

My definition of "under-confident" is "mean predicted proba < fraction of positives". (I do not think we define the notion of "confidence" in the example)

ogrisel replied:
This is what I see on https://github.com/scikit-learn/scikit-learn/pull/28231#issuecomment-1906650776 (the top plot, the one for this PR):

E.g. for mean prediction of 0.8, the frequency of the positively labeled observations for the LR model (green line) is above the diagonal (around 0.9).

The LR models predicts 0.8 instead of 0.9 (on average), hence is under-confident.

thomasjpfan replied:
If I just plot the calibration for logistic regression, I see two regions, one that is over-confident and another that is under-confident:

![Screenshot 2024-02-02 at 4 46 15 PM](https://github.com/scikit-learn/scikit-learn/assets/5402633/2990b50a-558f-4880-b9b6-55ee17ce5375)

Am I missing something here?

ogrisel replied:
Ah I see. I would have intuitively both sides as under-confident because on the left side, the models predicts the negative class (with the default decision threshold at 0.5), so in that respect it is under-confident everywhere to me (assuming the curve crosses the diagonal at 0.5).

Not sure what to do. Shall I remove the phrasing under-confident / over-confident completely?

ogrisel replied:
In 4fd1215, I defined under-confidence as making predictions too close to 0.5 compared to the fraction of the positive class. I think this is both concise, intuitive and accurate enough.

thomasjpfan replied:
Yea, it makes sense as long as we define what under-confidence means. Is there a reference that gives a standard definition for under-confidence in the context of calibration?

lorentzenchr replied:
What about just „over-estimate“ and „under-estimate“?

thomasjpfan commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,48 +156,84 @@ def predict_proba(self, X):
+# loglikelihood via inner cross-validation.
+#
+# The negative loglikelihood is a proper scoring rule for probabilistic
+# classification models. It means, that in the limit of infinite training data,
+# it is minimized by a model that predicts the true conditional probabilities
```
From here `it` is used a few times and can be hard to see that `it== negative loglikelihood

```suggestion
# classification models. In the limit of infinite training data, the
# negative log-likelihood is minimized by a model that predicts the true conditional probabilities
```

thomasjpfan commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,48 +156,84 @@ def predict_proba(self, X):
+# automatically tune the `C` regularization parameter to minimize the negative
+# loglikelihood via inner cross-validation.
+#
+# The negative loglikelihood is a proper scoring rule for probabilistic
```
```suggestion
# The negative log-likelihood is a proper scoring rule for probabilistic
```

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel committed:
Fix linting

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,48 +156,84 @@ def predict_proba(self, X):
+# :class:`~sklearn.linear_model.LogisticRegressionCV` returns reasonably well
+# calibrated predictions despite the small training set size: its reliability
+# curve is the closest to the diagonal. The reason is two-fold: this estimator
+# directly adjusts its trainable parameter to minimize negative loglikelihood
```
A likelihood
```suggestion
# directly adjusts its trainable parameter to minimize negative log-likelihood
```
A log-likelihood only makes sense with a distribution which we don‘t mention. Also, only very few likelihood functions can serve as proper scoring rules.
I would just always use the term log loss and not mention likelihood.

lorentzenchr replied:
This is the first reason and the argument would be better structured with mentioning proper scoring rule here not later.

ogrisel replied:
Alright let me see how I can reorg this.

ogrisel replied:
I pushed a new commit to reorganize the analysis: I introduce strictly proper scoring rules earlier, and I only use the log-loss name.

ogrisel committed:
Reorganized the beginning of the analysis section to address feedback from reviewers

ogrisel commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -149,62 +155,107 @@ def predict_proba(self, X):
-# :class:`~sklearn.linear_model.LogisticRegression` returns well calibrated
-# predictions as it directly optimizes log-loss. In contrast, the other methods
-# return biased probabilities, with different biases for each method:
+# %% Analysis of the results
+# -----------------------
```
```suggestion
# %% Analysis of the results
# --------------------------
```

ogrisel committed:
Fixed auto text wrap that broken the formatting of the references

ogrisel committed:
More formatting fix.

ogrisel committed:
Merge branch 'main' into improve-calibration-example

ogrisel committed:
One more formating fix and increase max_iter

ogrisel commented:
Thanks for the reviews. I think I addressed all the comments. The [rendering](https://output.circle-artifacts.com/output/job/4a1bcdbd-403e-41b0-885c-62f1c2055e17/artifacts/0/doc/auto_examples/calibration/plot_compare_calibration.html) looks good now.

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -26,8 +26,14 @@
+# remaining 16 are uninformative (random numbers).
+#
+# Of the 100,000 samples, 100 will be used for model fitting and the remaining
+# for testing. Note that this split is quite unusual: the goal is to obtain a
```
```suggestion
# for testing. Note that this split is quite unusual: the goal is to obtain
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,114 @@ def predict_proba(self, X):
+# curve is the closest to the diagonal among the four models.
+#
+# Logistic regression is trained by minimizing the log-loss which is a strictly
+# proper scoring rule: in the limit of infinite training data, strictly proper
```
We are getting closer;-)

I'm a bit undecided whether or not to mention asymptomatic arguments. We don’t need to, the expectation of proper scoring rules is minimized by the true probabilities. The asymptotic argument comes in only when estimating this expectation.

Note that also the RF uses proper scoring rules, e.g. Gini criterion is the squared error evaluated with the mean of that node as prediction.
I guess that it is more the overfitting difference that we see.

Arrrg, difficult to keep it simple.

ogrisel replied:
> I'm a bit undecided whether or not to mention asymptomatic arguments. We don’t need to, the expectation of proper scoring rules is minimized by the true probabilities. The asymptotic argument comes in only when estimating this expectation.

I think it's important because most of the time machine learning practitioners are interested in out-of-sample predictions.

> Note that also the RF uses proper scoring rules, e.g. Gini criterion is the squared error evaluated with the mean of that node as prediction.

Yes, I was thinking about updating the item about the RF but was worried to make it too complex.

> The fact that LR uses a canonical loss-link combination is also very important for the calibration results.

I can try to fit this somewhere, but since do not really expose custom link functions for classifiers, I am not sure if this will make the message too complex without justification in terms of choice of hyperparameters.

ogrisel replied:
I expanded the sections on RFs. I also conducted experiments on synthetic data. I confirm I can reproduce the effect reported by Caruana et al. (by playing with `max_features` and a dataset with a large number of uninformative features) so it's not just a theoretical artifact.

However, RF can also be strongly over-confident if the number of trees is not large enough to counteract the natural overfitting/over-confident behavior of a small number of unbounded trees. So the calibration picture can be really complex for RFs.

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,114 @@ def predict_proba(self, X):
+#   with this dataset, which contains 2 features generated as random linear
+#   combinations of the informative features. These correlated features are
+#   effectively being 'counted twice', resulting in pushing the predicted
+#   probabilities towards 0 and 1 [3]_. Note however that changing the seed
```
```suggestion
#   probabilities towards 0 and 1 [3]_. Note, however, that changing the seed
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,114 @@ def predict_proba(self, X):
+#
+# Feel free to re-run this example with different random seeds and other
+# dataset generation parameters to see how different the calibration plots can
+# look. In general, Logistic Regression and Random Forest will tend to be the
```
Maybe mention here that both use a proper scoring rule.

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel committed:
More detailed section about RF calibration

glemaitre commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,118 @@ def predict_proba(self, X):
+#
+# In this example the training set was intentionally kept very small. In this
+# setting, optimizing the log-loss can still lead to poorly calibrated models
+# because of overfitting. To mitigate this, the `LogisticRegressionCV` class
```
```suggestion
# because of overfitting. To mitigate this, the :class:`~sklearn.linear_model.LogisticRegressionCV` class
```

glemaitre commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,118 @@ def predict_proba(self, X):
+# look. In general, Logistic Regression and Random Forest will tend to be the
+# best calibrated classifiers, while SVC will often display the typical
+# under-confident miscalibration. The naive Bayes model is also often poorly
+# calibrated but the genaral shape of its calibration curve can vary widely
```
```suggestion
# calibrated but the general shape of its calibration curve can vary widely
```

glemaitre commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,118 @@ def predict_proba(self, X):
 #        naive Bayesian classifiers
 #        <https://citeseerx.ist.psu.edu/doc_view/pid/4f67a122ec3723f08ad5cbefecad119b432b3304>`_
 #        Zadrozny, Bianca, and Charles Elkan. Icml. Vol. 1. 2001.
+
+# %%
```
```suggestion
```

glemaitre reviewed (approved):
I like the improvements. I'm wondering if we should have a proper tutorial-based example to go more into details when it comes to models. I kind of like the to have the current example because it offers a 5-minutes to calibration example. But the topic is complex and it could be beneficial to go deeper in.

adrinjalali commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -86,17 +92,19 @@ def predict_proba(self, X):
 from sklearn.calibration import CalibrationDisplay
 from sklearn.ensemble import RandomForestClassifier
-from sklearn.linear_model import LogisticRegression
+from sklearn.linear_model import LogisticRegressionCV
 from sklearn.naive_bayes import GaussianNB
 
 # Create classifiers
-lr = LogisticRegression()
+lr = LogisticRegressionCV(
+    Cs=np.logspace(-6, 6, 101), cv=10, scoring="neg_log_loss", max_iter=1_000
+)
 gnb = GaussianNB()
 svc = NaivelyCalibratedLinearSVC(C=1.0, dual="auto")
-rfc = RandomForestClassifier()
+rfc = RandomForestClassifier(random_state=42)
```
As a user, it's confusing as why one model is doing a CV search and the others don't. If we need to do a search, shouldn't we be doing that for all models with their corresponding search spaces?

ogrisel replied:
I agree that ideally, all methods should have their hyperparameters tuned with a proper scoring rule.

It's very easy (and quite efficient) to do it for Logistic Regression while would require more efforts for the other methods and that would render the example much slower and verbose.

I could add an inline comment to say so.

ogrisel replied:
Also note: tuning C for the naively adapted SVC model would not fix its calibration (I tried).

ogrisel commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,61 +158,118 @@ def predict_proba(self, X):
+# under-confident: the predicted probabilities are a bit too close to 0.5
+# compared to the true fraction of positive samples.
+#
+# The other methods all output worse-calibrated probabilities:
```
```suggestion
# The other methods all output less well calibrated probabilities:
```

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel committed:
Wrap paragraph

ogrisel commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -86,17 +92,19 @@ def predict_proba(self, X):
+from sklearn.linear_model import LogisticRegressionCV
 from sklearn.naive_bayes import GaussianNB
 
 # Create classifiers
```
```suggestion
# Create classifiers: here we tune the main hyper-parameter automatically for
# the logistic regression model because its efficient and concise to do so.
# Ideally, we should run an hyper-parameter search for all the models but we
# don't do it here for the sake of keeping the example code concise and fast
# to execute.
```

ogrisel committed:
Add an inline comment about hparam tuning.

adrinjalali commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -86,17 +92,23 @@ def predict_proba(self, X):
-lr = LogisticRegression()
+# Create classifiers: here we tune the main hyper-parameter automatically for
+# the logistic regression model because its efficient and concise to do so.
+# Ideally, we should run an hyper-parameter search for all the models but we
```
```suggestion
# Ideally, we should run a hyper-parameter search for all the models but we
```

adrinjalali reviewed (approved):
Just a nit. Thanks Olivier.

ogrisel committed:
Update examples/calibration/plot_compare_calibration.py

Co-authored-by: Adrin Jalali <adrin.jalali@gmail.com>

ogrisel commented:
In retrospect I agree with @adrinjalali that it would be relevant to study the impact of tuning the parameters for the other models.

In particular, adjusting the smoothing for the Gaussian NB model can make it be as well calibrated as Logistic Regression.

For RFs, adjusting the number of trees and leaf size can make from over confident (few over fitting trees) to slightly under confident. However the RF parametrization with the best logloss has a very similar calibration curve to the one currently displayed in this example.

The SVC is always badly calibrated, whatever the regularization and that is expected because of the naive minmax scaling of the decision function values into the [0-1] range.

Still, this would be a lot of work: instead of doing a single joint plot with 4 model at once, I would be in favor of doing a more tutorial-style example that studies one model class at a time (with one or more plot per model class) and analyse the impact of the most important hyper-parameters on calibration and logloss.

Therefore am in favor of merging the current state of this PR as is, so as to be able to better back the arguments developed in #28171 and incrementally improve our message on model calibration. And later consider refactoring the 2 examples on model calibration to make them more tutorial-ish.

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -86,17 +92,23 @@ def predict_proba(self, X):
-# Create classifiers
-lr = LogisticRegression()
+# Create classifiers: here we tune the main hyper-parameter automatically for
+# the logistic regression model because its efficient and concise to do so.
```
This sentence doesn't work out

ogrisel replied:
I reworked it in the subsequent commits.

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -86,17 +92,23 @@ def predict_proba(self, X):
-lr = LogisticRegression()
+# Create classifiers: here we tune the main hyper-parameter automatically for
+# the logistic regression model because its efficient and concise to do so.
+# Ideally, we should run a hyper-parameter search for all the models but we
```
```suggestion
# For a fair comparison, we should run a hyper-parameter search for all the models but we
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
+# Logistic regression is trained by minimizing the log-loss which is a strictly
+# proper scoring rule: in the limit of infinite training data, strictly proper
+# scoring rules are minimized by the model that predicts the true conditional
+# probabilities and would therefore be perfectly calibrated. However, using a
```
```suggestion
# probabilities. That (hypothetical) model would therefore be perfectly calibrated. However, using a
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
+# probabilities and would therefore be perfectly calibrated. However, using a
+# proper scoring rule as training objective is not sufficient to guarantee a
+# well-calibrated model by itself: even with a very large training set,
+# logistic regression could still be poorly calibrated, if it was too
```
```suggestion
# logistic regression could still be poorly calibrated, if it was too strongly
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
+# proper scoring rule as training objective is not sufficient to guarantee a
+# well-calibrated model by itself: even with a very large training set,
+# logistic regression could still be poorly calibrated, if it was too
+# regularized or if the choice of input features made this model mis-specified
```
```suggestion
# regularized or if the choice and preprocessing of input features made this model mis-specified
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
+# well-calibrated model by itself: even with a very large training set,
+# logistic regression could still be poorly calibrated, if it was too
+# regularized or if the choice of input features made this model mis-specified
+# (e.g. the true decision boundary is a highly non-linear function of the input
```
```suggestion
# (Note that the true decision boundary of our dataset is a highly non-linear function of the input
```
or
```suggestion
# (e.g. if the true decision boundary of the dataset is a highly non-linear function of the input
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
-#   :class:`~sklearn.ensemble.RandomForestClassifier`, which is typical for
-#   maximum-margin methods [1]_ as they focus on difficult to classify samples
-#   that are close to the decision boundary (the support vectors).
+#   independent holds [2]_. However, features can be correlated and is the case
```
```suggestion
#   independent holds [2]_. However, features can be correlated and this is the case
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
+#   diagonal). This under-confident predictions are typical for maximum-margin
+#   methods [1]_.
+#
+# * :class:`~sklearn.ensemble.RandomForestClassifier`'s prediction histograms
+#   show peaks at approx. 0.2 and 0.9 probability, while probabilities close to
```
```suggestion
# * :class:`~sklearn.ensemble.RandomForestClassifier`'s prediction histogram
#   shows peaks at approx. 0.2 and 0.9 probability, while probabilities close to
```

lorentzenchr commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -150,60 +162,116 @@ def predict_proba(self, X):
+#
+# * :class:`~sklearn.ensemble.RandomForestClassifier`'s prediction histograms
+#   show peaks at approx. 0.2 and 0.9 probability, while probabilities close to
+#   0 or 1 are very rare. An explanation for this is given by Niculescu-Mizil
+#   and Caruana [1]_: "Methods such as bagging and random forests that average
```
```suggestion
#   0 or 1 are very rare. An explanation for this is given by [1]_:
#   "Methods such as bagging and random forests that average
```

ogrisel commented on examples/calibration/plot_compare_calibration.py:
```diff
@@ -86,17 +92,23 @@ def predict_proba(self, X):
 
-# Create classifiers
-lr = LogisticRegression()
+# Create classifiers: here we tune the main hyper-parameter automatically for
+# the logistic regression model because its efficient and concise to do so.
```
```suggestion
# Define the classifiers to be compared in the study. Note that we use a variant of the
# logistic regression model that can internally tuned its regularization parameter
# efficient and concise to do so.
```

ogrisel committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

ogrisel committed:
Wrap paragraph + grammar fix

ogrisel commented:
@lorentzenchr I think I addressed all the points of your last review.

lorentzenchr commented:
Those doc updates often take longer than anticipated :smirk:

glemaitre commented:
I'll backport this one
